### PR TITLE
Fix traefik path routing

### PR DESCRIPTION
**Pull request by arturo-seijas**
Applicable spec: <link> N/A

### Overview

Set a prefix for Jenkins when path routing is enabled. This means that the api calls will also need to take that prefix into account.

In order to achieve this, the prefix will be passed to the `jenkins.Jenkins` instance when constructed and the methods that ultimately make api calls are converted into instance methods

### Rationale

N/A

### Juju Events Changes

N/A

### Module Changes

Spreads across all modules

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

**Comment by amandahla** [comment]
Is not necessary to observe the event _on_ingress_ready to get the URL when ready?

File: src/jenkins.py [hunk 1]
```diff
@@ -144,6 +146,467 @@ def _get_api_credentials(container: ops.Container) -> Credentials:
...
+    @property
+    def login_url(self) -> str:
+        """Get the Jenkins login URL.
+
+        Returns: the login URL.
+        """
+        return f"{self.web_url}{LOGIN_PATH}"
+
+    @property
+    def version(self) -> str:
```
Comment (Thanhphan1147) [inline src/jenkins.py:187]: Since we only expect the server version to change during a charm upgrade, should we just use a constant here?

**Review by Thanhphan1147 — COMMENTED** [review]
only a minor comment, I haven't looked at tests yet

File: src/ingress.py [hunk 2]
```diff
@@ -28,3 +30,16 @@ def __init__(self, charm: ops.CharmBase, key: str, relation_name: str):
             port=jenkins.WEB_PORT,
             strip_prefix=True,
         )
+
```
Comment (Thanhphan1147) [inline src/ingress.py:33]: should we add an integration test for this? (traefik in path routing mode)

File: tox.ini [hunk 3]
```diff
@@ -125,10 +125,11 @@ deps =
...
     pytest --tb native --ignore={[vars]tst_path}unit --log-cli-level=INFO -s {posargs}
-    ; uncomment the following lines to use with output of tox -e build.
-    ; --jenkins-image=localhost:32000/jenkins:test \
-    ; --charm-file=./jenkins-k8s_ubuntu-22.04-amd64.charm \
-    ; --kube-config=~/.kube/config
+    # uncomment the following lines to use with output of tox -e build.
+    # --jenkins-image=localhost:32000/jenkins:test \
+    # --charm-file=./jenkins-k8s_ubuntu-22.04-amd64.charm \
+    # --kube-config=~/.kube/config \
+    # --keep-models
```
Comment (Thanhphan1147) [inline tox.ini:132]: I don't think we'd want to keep the test models by default right?

File: tests/unit/test_charm.py [hunk 4]
```diff
@@ -125,67 +126,66 @@ def test__on_jenkins_pebble_ready_get_version_error(
     monkeypatch: pytest.MonkeyPatch,
 ):
     """
-    arrange: given a patched jenkins.get_version function that raises an exception.
+    arrange: given a patched jenkins.version function that raises an exception.
```
Comment (Thanhphan1147) [inline tests/unit/test_charm.py:129]:
```suggestion
    arrange: given a patched jenkins.version property that raises an exception.
```

**Review by Thanhphan1147 — DISMISSED** [review]
LGTM, only a few minor comments

**Review by weiiwang01 — DISMISSED** [review]

File: src-docs/auth_proxy.py.md
Comment (jdkandersson) [inline src-docs/auth_proxy.py.md:1]: Should this file be here?

File: src/charm.py [hunk 5]
```diff
@@ -5,6 +5,8 @@
 
 """Charm Jenkins."""
 
+# pylint: disable=too-many-instance-attributes
```
Comment (jdkandersson) [inline src/charm.py:8]: Should this include a note as to why it is reasonable to disable this?

File: src/charm.py [hunk 6]
```diff
@@ -99,14 +101,25 @@ def _get_pebble_layer(self, jenkins_env: jenkins.Environment) -> ops.pebble.Laye
...
         return ops.pebble.Layer(layer)
 
+    def calculate_env(self) -> jenkins.Environment:
+        """Return a dictionary for Jenkins Pebble layer.
+
+        Returns:
+            The dictionary mapping of environment variables for the Jenkins service.
+        """
+        return jenkins.Environment(
+            JENKINS_HOME=str(state.JENKINS_HOME_PATH),
```
Comment (jdkandersson) [inline src/charm.py:119]: does this need to be on the state or could it just be a constant?

File: src/jenkins.py [hunk 7]
```diff
@@ -3,6 +3,8 @@
 
 """Functions to operate Jenkins."""
 
+# pylint: disable=too-many-lines
```
Comment (jdkandersson) [inline src/jenkins.py:6]: Perhaps this needs to be split into a few submodules now?

File: src/jenkins.py [hunk 8]
```diff
@@ -144,6 +146,467 @@ def _get_api_credentials(container: ops.Container) -> Credentials:
         raise JenkinsBootstrapError("Admin API credentials not yet setup.") from exc
 
 
+class Jenkins:
```
Comment (jdkandersson) [inline src/jenkins.py:149]: What is the reason please for changing this to be a class?

File: src/jenkins.py [hunk 9]
```diff
@@ -144,6 +146,467 @@ def _get_api_credentials(container: ops.Container) -> Credentials:
...
+        """
+        self.environment = environment
+
+    @property
+    def web_url(self) -> str:
+        """Get the Jenkins web URL.
+
+        Returns: the web URL.
+        """
+        env_dict = typing.cast(typing.Dict[str, str], self.environment)
```
Comment (jdkandersson) [inline src/jenkins.py:175]: Isn't this already declared as a dict? What is the reason for needing the cast here please?

**Review by jdkandersson — COMMENTED** [review]
have reviewed up to the tests

File: src/ingress.py (hunk 2 above)
Comment (arturo-seijas) [reply src/ingress.py:33]: I switched traefik's routing mode to `route` in the corresponding fixture. That should suffice (if route works, subdomain will work too)
```
traefik = await model.deploy(
        "traefik-k8s",
        channel="edge",
        trust=True,
        config={
            "external_hostname": external_hostname,
            "routing_mode": "path",
            "enable_experimental_forward_auth": True,
        },
    )
  
   ```

File: src/charm.py (hunk 6 above)
Comment (arturo-seijas) [reply src/charm.py:119]: It's a constant that is defined in the `state` module. I'll move it to `jenkins`

File: src/jenkins.py (hunk 9 above)
Comment (arturo-seijas) [reply src/jenkins.py:175]: Added a comment explaining why

File: src/jenkins.py (hunk 7 above)
Comment (arturo-seijas) [reply src/jenkins.py:6]: Yes, but it seems the me the PR is already too bit. I plan to do that in a follow up PR along with some more refactoring

File: src/charm.py (hunk 5 above)
Comment (arturo-seijas) [reply src/charm.py]: Removed. It's not needed anymore

File: src/jenkins.py (hunk 8 above)
Comment (arturo-seijas) [reply src/jenkins.py:147]: The client and all the method that use it are dependent on the prefix that is set in the pebble plan, as such, the easiest way to track that prefix is to set that prefix as an attribute of the `jenkins` class

File: src-docs/auth_proxy.py.md
Comment (arturo-seijas) [reply src-docs/auth_proxy.py.md:1]: Removed

File: src/jenkins.py (hunk 1 above)
Comment (arturo-seijas) [reply src/jenkins.py:185]: That would mean having the version defined in two places, the rockfile and the charm code. Anyway, this was already the existing behaviour, so I'd rather keep it as it is at least in this PR

**Review by weiiwang01 — APPROVED** [review]

**Review by Thanhphan1147 — APPROVED** [review]

**Review by weiiwang01 — APPROVED** [review]

**Review by Thanhphan1147 — APPROVED** [review]

**Review by mthaddon — APPROVED** [review]
LGTM, thx

**Pull request merged**